### PR TITLE
retention: Drop prefix/checksum code if not used

### PR DESCRIPTION
**Pull request by nordicjm**
Will exclude the code for prefix and checksum functionality if none of the entries on a device make use of these features, saves ~816 bytes when both features are excluded.

### CODE REVIEW

**Review by de-nordic — APPROVED** [review]

**Review by Laczen — DISMISSED** [review]
This is an example of how development should not be done:

1. Either these things are required and they should be inside the code, or they are not and then they should be removed.
2. A simpler system should not be made by configuration options/descriptions in the dts, this makes testing very difficult and error prone. Instead a system should be simple from the start and add complexity if required, this allows separate testing of the simple and complex systems.

I have been warning for this from the introduction of retention.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by henrikbrixandersen** [comment]
@Laczen This may be your thoughts on this (others may have had a need for this aded functionality). However, I do not believe this is a valid reason for blocking this PR.

**Comment by Laczen** [comment]
> @Laczen This may be your thoughts on this (others may have had a need for this aded functionality). However, I do not believe this is a valid reason for blocking this PR.

@henrikbrixandersen if the "need" for this PR is a comparison please do not add this, when the comparison is done this code will live without purpose and confuse users. If there is a real "need" the user should have been directed to another subsystem.

**Comment by henrikbrixandersen** [comment]
> @henrikbrixandersen if the "need" for this PR is a comparison please do not add this, when the comparison is done this code will live without purpose and confuse users. If there is a real "need" the user should have been directed to another subsystem.

I am trying hard to follow your argumentation here, but I still don't get it. 

This PR is a code optimization to only compile-in support for prefix and/or checksums if those features are enabled via devicetree.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@Laczen  I also do not see your point.

> If there is a real "need" the user should have been directed to another subsystem.

what does this mean exactly? This is the subsystem for retention, we do not have anything else and this is an optimization, that does not involve user at all and all happens under the hood.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
> d. No prefix or crc are present: the validity check makes no sense and instead the user should be guided into using retained_mem directly. The system should not respond that the data is valid.

`retained_mem` does not support partitioning @Laczen, so I think it makes sense to allow non-CRCed partitions to use the retention subsystem.

**Comment by Laczen** [comment]
> > d. No prefix or crc are present: the validity check makes no sense and instead the user should be guided into using retained_mem directly. The system should not respond that the data is valid.
> 
> `retained_mem` does not support partitioning @Laczen, so I think it makes sense to allow non-CRCed partitions to use the retention subsystem.

OK, this is what happens when you are trying to add 2 functionalities in one system @carlescufi . Probably the partitioning should not be done in `retention` but in `retained_mem` instead.

**Comment by nordicjm** [comment]
> > > d. No prefix or crc are present: the validity check makes no sense and instead the user should be guided into using retained_mem directly. The system should not respond that the data is valid.
> > 
> > 
> > `retained_mem` does not support partitioning @Laczen, so I think it makes sense to allow non-CRCed partitions to use the retention subsystem.
> 
> OK, this is what happens when you are trying to add 2 functionalities in one system @carlescufi . Probably the partitioning should not be done in `retention` but in `retained_mem` instead.

No, the partitioning is done in retention which is the correct place for it because it is a software feature - this was decided collectively when retained memory was brought up at the architecture working group. There is nothing wrong with having multiple types of configuration in a system, I can't really understand what you are complaining about or any merit in it, this would be like complaining about the MCUmgr subsystem because it allows you to have a transport over UART and bluetooth, and also because you can adjust the buffer sizes for those transports, and you can also turn what included commands you will have, what's wrong with configuration?

**Comment by Laczen** [comment]
> > > > d. No prefix or crc are present: the validity check makes no sense and instead the user should be guided into using retained_mem directly. The system should not respond that the data is valid.
> > > 
> > > 
> > > `retained_mem` does not support partitioning @Laczen, so I think it makes sense to allow non-CRCed partitions to use the retention subsystem.
> > 
> > 
> > OK, this is what happens when you are trying to add 2 functionalities in one system @carlescufi . Probably the partitioning should not be done in `retention` but in `retained_mem` instead.
> 
> No, the partitioning is done in retention which is the correct place for it because it is a software feature - this was decided collectively when retained memory was brought up at the architecture working group. There is nothing wrong with having multiple types of configuration in a system, I can't really understand what you are complaining about or any merit in it, this would be like complaining about the MCUmgr subsystem because it allows you to have a transport over UART and bluetooth, and also because you can adjust the buffer sizes for those transports, and you can also turn what included commands you will have, what's wrong with configuration?

The problem is that the PR breaks the api. It needs changes in the case that only a prefix is used (clearing the prefix, writing, setting the prefix) and it needs to generate a not-supported when both the prefix and the crc have been disabled.

The problem that partitioning is done together with a validity check can be solved by instead of directly using retained_mem use boot_info (that provides just the partitioning). Boot_info can be modified to also support using retained_mem as its backend (instead of ram over the flash_simulator). The reason it doesn't do so is that I find the name retained_mem misleading for the memory region backend. There might be a solution for this if it is known what the state of the memory is when it is cleared by the cpu/os.

**Comment by nordicjm** [comment]
> The problem is that the PR breaks the api. It needs changes in the case that only a prefix is used (clearing the prefix, writing, setting the prefix) and it needs to generate a not-supported when both the prefix and the crc have been disabled.

No, this PR does not change the API.

**Comment by Laczen** [comment]
> > The problem is that the PR breaks the api. It needs changes in the case that only a prefix is used (clearing the prefix, writing, setting the prefix) and it needs to generate a not-supported when both the prefix and the crc have been disabled.
> 
> No, this PR does not change the API.

You are right, it doesn't: the api is reporting -ENOTSUP when both the crc and the prefix are disabled. This of course moves the responsibility of checking this to the user of the subsystem. What should the user do when it gets a not-supported as result of 
`is_valid()` it receives a `-ENOTSUP` ?
 
The support of only a prefix needs modification as it is needed to be set to a bad value before performing a write.

**Comment by nordicjm** [comment]
The user can do whatever they want to do with it, the information is provided and the user can make a decision. Someone storing a reboot count is going to do things differently than someone storing a one time certificate which controls a highly important mission critical system.

> The support of only a prefix needs modification as it is needed to be set to a bad value before performing a write.

Have you tried submitting a PR?

**Comment by Laczen** [comment]
> The user can do whatever they want to do with it, the information is provided and the user can make a decision. Someone storing a reboot count is going to do things differently than someone storing a one time certificate which controls a highly important mission critical system.
> 

I have always seen this system as something where the crc is obligatory and the prefix would be optional. This makes sense because the API is providing a `is_valid()` function. The `is_valid()` reporting -ENOTSUP would then be more of a config error.

If the intention of this system is to be used with a crc there is no need to add code to compile it out.

**Comment by carlescufi** [comment]
@Laczen I really do not think blocking this PR makes sense

> I have always seen this system as something where the crc is obligatory and the prefix would be optional. This makes sense because the API is providing a `is_valid()` function. The `is_valid()` reporting -ENOTSUP would then be more of a config error.

This subsystem provides the functionality to **optionally** add a header and/or CRC to the retained data. That functionality is, again, **entirely optional** but still part of the subsystem. When you do not use headers and/or CRCs then the `is_valid()` function is obviously not applicable and does nothing. But this is the case in many other subsystems and APIs.

> If the intention of this system is to be used with a crc there is no need to add code to compile it out.

Again, this subsystem can be used with or without a CRC, hence this fix.

Please reconsider blocking this PR.
